### PR TITLE
✨ Add maze order (1 or 0) corresponding to the maze density

### PR DESCRIPTION
**Pull request by bal7hazar**
## Introduced changes

Add a new argument for maze and corridor generations, the density order.
The order of the maze, it must be 0 or 1, the higher the order the less dense the maze will be.

Note: I didnt managed higher orders since the definition is ambiguous and the scaling could be a problem.

Here an example:

**Order 0** (current implementation)
```rust
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 1 0 1 1 1 0 1 1 1 1 1 1 1 1 0 1 0
0 1 1 0 1 0 1 1 0 0 1 0 0 0 1 1 1 0
0 0 1 1 1 1 0 1 1 0 1 1 1 1 0 0 1 0
0 1 1 0 0 1 0 0 1 1 0 1 0 1 0 1 0 0
0 1 0 0 0 1 1 1 1 0 1 0 0 1 1 1 1 0
0 1 1 0 1 0 0 0 0 0 1 1 1 0 1 0 1 0
0 0 1 1 1 0 1 1 1 1 0 1 0 1 0 1 1 0
0 1 0 1 0 1 1 0 0 1 1 1 0 1 1 1 0 0
0 1 0 1 1 1 0 1 1 1 0 1 1 0 1 0 1 0
0 1 1 0 0 0 1 0 1 0 1 1 0 1 1 1 1 0
0 0 1 1 1 1 1 1 0 1 1 0 0 1 0 1 0 0
0 1 1 0 1 0 0 1 1 1 0 1 1 1 0 1 1 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
```
**Order 1** (new implementation, with the same seed)
```rust
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
0 1 1 1 1 1 0 0 1 1 1 1 1 1 1 1 0 0
0 0 1 0 0 1 0 1 1 0 1 0 0 0 0 1 1 0
0 1 1 0 0 0 0 1 0 0 1 1 1 1 0 0 1 0
0 0 1 1 1 1 0 1 1 0 0 0 0 1 0 0 0 0
0 1 1 0 0 1 0 0 1 0 1 1 0 1 1 1 1 0
0 1 0 0 1 1 1 1 1 1 1 0 0 0 0 0 1 0
0 1 1 0 0 0 0 0 0 0 0 0 1 0 0 1 1 0
0 0 1 1 0 0 1 1 1 1 0 1 1 0 1 1 0 0
0 0 0 1 0 1 1 0 0 1 1 1 0 0 1 0 0 0
0 1 0 1 1 1 0 0 0 1 0 1 0 1 1 1 1 0
0 1 0 0 0 0 0 1 0 1 0 0 0 1 0 0 1 0
0 1 1 1 1 1 1 1 1 1 1 0 1 1 0 1 1 0
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
```

Basically, order 1 disallow these patterns (having 2 floors connected diagonally):

```rust
1 1 0
0 0 1
0 1 1
```

### CODE REVIEW

File: crates/map/src/helpers/mazer.cairo [hunk 1]
```diff
@@ -86,38 +101,53 @@ pub impl Mazer of MazerTrait {
...
+    fn check(
+        maze: felt252, width: u8, height: u8, order: u8, position: u8, direction: Direction
+    ) -> bool {
+        // [Check] Order is valid
+        assert(order <= 1, errors::MAZER_INVALID_ORDER);
+        // [Check] Check the position
         let (x, y) = (position % width, position / width);
         match direction {
-            Direction::North => (y <= height - 3)
+            Direction::North => (y < height - 2)
```
Comment (glihm) [inline crates/map/src/helpers/mazer.cairo:119]: the `-2` here looks a bit magic, perhaps having this in a constant that have a more descriptive name could help the understanding.

Comment (glihm) [reply crates/map/src/helpers/mazer.cairo:119]: Same for some offset used in this function.

Comment (bal7hazar) [reply crates/map/src/helpers/mazer.cairo:119]: Actually 3 was weird but size - 2 is actually the range of the indexes (from 0 to size - 1) substracting 1 to ensure we didnt reached the edge yet.

It could be rewrite such as:

```cairo
last_index = height - 1;
one_before_last_index = last_index - 1;
match direction {
    Direction::North => (y <= one_before_last_index)
...
}
```

But maybe I could just add some comments to explain it, wdyt?

Comment (glihm) [reply crates/map/src/helpers/mazer.cairo:119]: Comment would be fine yeah, was trying to grasp the idea of those indexes. Thanks!
